### ISSUE TITLE
Develop Model Prototype

### ISSUE THREAD

**Issue by jonah-chen** [jonah-chen/midi-shark#3]
## Tips:
- If training/testing/debugging on the GPU, use the line 
```py
torch.set_default_tensor_type('torch.cuda.FloatTensor')
```
to set create tensors on the GPU by default. DO NOT USE `.cuda()` function unless there is a very good reason. If strictly necessary, justify with comments.

## My Proposed Approach

If you have any suggestions, please let us know in a comment.

In my opinion, we should start with developing the most theoretical parts of the model because this would help solidify understanding of the various methods we are implementing. After we complete this, we can integrate everything together into the model. I have broken it down into several components.  

### Transformer (Decoder for onset and frame heads):

Single components:
- Positional Encoding
- Generating Query/Key/Value (s)
- Simple Attention
- Multi-head Attention
- Add & Norm

More complex networks:
- Input and output embedding
- N layer encoder structure
- M layer decoder structure
- Prediction head(s)

### Conv Stack/Image Segmentation (Encoder for image segmentation model)

- Conv Stack in 2016 paper
- Adapted Deeplab v3 (I have implemented vanilla deeplabv3 in tensorflow before, but we need to adapt the ideas from the paper to suit our needs, also write it in pytorch)
- Try other segmentation models

### Full Model

- Answer how the data will flow throughout these components (and simpler components like sigmoids, fully-connected layers, softmax, etc).
- Experiment with velocity and other useful information we can extract??

P.S. I never used github properly before so I hope this is what I'm supposed to do.
